core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooAclUtil.java

```diff
@@ -0,0 +1,166 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.fate.zookeeper;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import org.apache.zookeeper.ZooDefs;
+import org.apache.zookeeper.data.ACL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ZooAclUtil {
+
+  private static final Logger log = LoggerFactory.getLogger(ZooAclUtil.class);
+
+  /**
+   * Utility class - prevent instantiation.
+   */
+  private ZooAclUtil() {}
+
+  public static String extractAuthName(ACL acl) {
+    requireNonNull(acl, "provided ACL cannot be null");
+    try {
+      return acl.getId().getId().trim().split(":")[0];
+    } catch (Exception ex) {
+      log.debug("Invalid ACL passed, cannot parse id from '{}'", acl);
+      return "";
+    }
+  }
+
+  /**
+   * translate the ZooKeeper ACL perm bits into a string. The output order is cdrwa (when all perms
+   * are set)
+   *
+   * Mirrors `org.apache.zookeeper.ZKUtil.getPermString()` added in more recent ZooKeeper versions.
+   */
+  public static String translateZooPerm(final int perm) {
+    if (perm == ZooDefs.Perms.ALL) {
+      return "cdrwa";
+    }
+    StringBuilder sb = new StringBuilder();
+    if ((perm & ZooDefs.Perms.CREATE) != 0) {
+      sb.append("c");
+    }
+    if ((perm & ZooDefs.Perms.DELETE) != 0) {
+      sb.append("d");
+    }
+    if ((perm & ZooDefs.Perms.READ) != 0) {
+      sb.append("r");
+    }
+    if ((perm & ZooDefs.Perms.WRITE) != 0) {
+      sb.append("w");
+    }
+    if ((perm & ZooDefs.Perms.ADMIN) != 0) {
+      sb.append("a");
+    }
+    if (sb.length() == 0) {
+      return "invalid";
+    }
+    return sb.toString();
+  }
+
+  /**
+   * Process the ZooKeeper acls and return a structure that shows if the accumulo user has ALL
+   * (cdrwa) access and if any other user has anything other than read access.
+   * <ul>
+   * <li>Accumulo having access is expected - anything else needs checked
+   * <li>Anyone having more than read access is unexpected and should be checked.
+   * </ul>
+   *
+   * @param acls ZooKeeper ACLs for a node
+   * @return acl status with accumulo and other accesses.
+   */
+  public static ZkAccumuloAclStatus checkWritableAuth(List<ACL> acls) {
+
+    ZkAccumuloAclStatus result = new ZkAccumuloAclStatus();
+
+    for (ACL a : acls) {
+      String name = extractAuthName(a);
+      // check accumulo has or inherits all access
+      if ("accumulo".equals(name) || "anyone".equals(name)) {
+        if (a.getPerms() == ZooDefs.Perms.ALL) {
+          result.setAccumuloHasFull();
+        }
+      }
+      // check if not accumulo and any perm bit other than read is set
+      if (!"accumulo".equals(name) && ((a.getPerms() & ~ZooDefs.Perms.READ) != 0)) {
```

Comment:
wouldn't you want to check that "anyone" matches `name`? There could be another user that has an ACL on this node.

Reply:
This is checking that if it is not the accumulo user. then if other users can do anything other than read - it could be the generic ZooKeeper `anyone` of maybe a custom other - if any user other than accumulo has an access other than read they "could" modify the node - so it sets the any may update flag - it was intended to be access by non-accumulo. anyone or any other zk user.

Reply:
ok, I misunderstood, maybe a better name would be `othersCanUpdate` ?

Reply:
use of `anyone` here implies `world` instead of a different user.